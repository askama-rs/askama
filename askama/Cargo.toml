[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
