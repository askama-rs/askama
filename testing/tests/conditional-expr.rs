use askama::Template;

#[test]
fn test_associativity() {
    // Here we test the associativity of conditional expressions.
    // `a if b else c if d else e` is the same as `a if b else (c if d else e)` in Python,
    // so we want to parse it right-to-left, too.

    #[derive(Template)]
    #[template(
        source = r#"
            {{-
                "a" if a else
                "b" if b else
                "c" if c else
                "d" if d else
                "-"
            -}}
        "#,
        ext = "html"
    )]
    struct Abcd<'a> {
        a: bool,
        b: &'a bool,
        c: &'a &'a bool,
        d: &'a &'a &'a bool,
    }

    for a in [false, true] {
        for b in [&false, &true] {
            for c in [&&false, &&true] {
                for d in [&&&false, &&&true] {
                    let expected = match (a, b, c, d) {
                        (true, _, _, _) => "a",
                        (_, true, _, _) => "b",
                        (_, _, true, _) => "c",
                        (_, _, _, true) => "d",
                        _ => "-",
                    };
                    let actual = Abcd { a, b, c, d }.to_string();
                    assert_eq!(actual, expected, "a={a} b={b} c={c} d={d}");
                }
            }
        }
    }
}

#[test]
fn test_associativity_reversed() {
    // This test makes sure explicit grouping works.

    #[derive(Template)]
    #[template(
        source = r#"
            {{-
                (
                    (
                        (
                            "-" if !d else "d"
                        ) if !c else "c"
                    ) if !b else "b"
                ) if !a else "a"
            -}}
        "#,
        ext = "html"
    )]
    struct Abcd<'a> {
        a: bool,
        b: &'a bool,
        c: &'a &'a bool,
        d: &'a &'a &'a bool,
    }

    for a in [false, true] {
        for b in [&false, &true] {
            for c in [&&false, &&true] {
                for d in [&&&false, &&&true] {
                    let expected = match (a, b, c, d) {
                        (true, _, _, _) => "a",
                        (_, true, _, _) => "b",
                        (_, _, true, _) => "c",
                        (_, _, _, true) => "d",
                        _ => "-",
                    };
                    let actual = Abcd { a, b, c, d }.to_string();
                    assert_eq!(actual, expected, "a={a} b={b} c={c} d={d}");
                }
            }
        }
    }
}

#[test]
fn test_associativity_infix() {
    // Being a ternary operator, a conditional-expression can be the condition of a
    // conditional-expression without the need for explicit grouping.

    #[derive(Template)]
    #[template(
        source = r#"{{- "a" if then if cond else otherwise else "b" -}}"#,
        ext = "html"
    )]
    struct ThenOtherwise<'a> {
        cond: bool,
        then: &'a bool,
        otherwise: &'a &'a bool,
    }

    for cond in [false, true] {
        for then in [&false, &true] {
            for otherwise in [&&false, &&true] {
                let expected = if if cond { *then } else { **otherwise } {
                    "a"
                } else {
                    "b"
                };
                let actual = ThenOtherwise {
                    cond,
                    then,
                    otherwise,
                }
                .to_string();
                assert_eq!(
                    actual, expected,
                    "cond={cond} then={then} otherwise={otherwise}"
                );
            }
        }
    }
}

// Conditional expressions are just expressions, so you should be able to use them in all
// places that expect an expression.

#[test]
fn test_call_argument() {
    #[derive(Template)]
    #[template(
        source = r#"{{- method(then if cond else otherwise) -}}"#,
        ext = "html"
    )]
    struct AsArgument {
        cond: bool,
        then: i32,
        otherwise: i32,
    }

    impl AsArgument {
        fn method(&self, value: &i32) -> i32 {
            *value
        }
    }

    assert_eq!(
        AsArgument {
            cond: true,
            then: 1,
            otherwise: -1
        }
        .to_string(),
        "1"
    );
    assert_eq!(
        AsArgument {
            cond: false,
            then: 1,
            otherwise: -1
        }
        .to_string(),
        "-1"
    );
}

#[test]
fn test_filter_argument() {
    #[derive(Template)]
    #[template(
        source = r#"{{- input | add(then if cond else otherwise) -}}"#,
        ext = "html"
    )]
    struct AsFilterArgument {
        input: i32,
        cond: bool,
        then: i32,
        otherwise: i32,
    }

    mod filters {
        pub(crate) fn add(input: &i32, _: &dyn askama::Values, arg: &i32) -> askama::Result<i32> {
            Ok(*input + *arg)
        }
    }

    assert_eq!(
        AsFilterArgument {
            input: 10,
            cond: true,
            then: 1,
            otherwise: -1
        }
        .to_string(),
        "11"
    );
    assert_eq!(
        AsFilterArgument {
            input: 10,
            cond: false,
            then: 1,
            otherwise: -1
        }
        .to_string(),
        "9"
    );

    for input in -10..10 {
        for cond in [false, true] {
            for then in -10..10 {
                for otherwise in -10..10 {
                    let expected = if cond {
                        input + then
                    } else {
                        input + otherwise
                    };
                    let actual = AsFilterArgument {
                        input,
                        cond,
                        then,
                        otherwise,
                    };
                    assert_eq!(expected.to_string(), actual.to_string());
                }
            }
        }
    }
}

#[test]
fn test_if_condition() {
    #[derive(Template)]
    #[template(
        source = r#"
            {%- if then if cond else otherwise -%}
                a
            {%- else -%}
                b
            {%- endif -%}
        "#,
        ext = "html"
    )]
    struct AsIfCondition {
        cond: bool,
        then: bool,
        otherwise: bool,
    }

    for cond in [false, true] {
        for then in [false, true] {
            for otherwise in [false, true] {
                let expected = if if cond { then } else { otherwise } {
                    "a"
                } else {
                    "b"
                };
                let actual = AsIfCondition {
                    cond,
                    then,
                    otherwise,
                }
                .to_string();
                assert_eq!(
                    actual, expected,
                    "cond={cond} then={then} otherwise={otherwise}"
                );
            }
        }
    }
}
