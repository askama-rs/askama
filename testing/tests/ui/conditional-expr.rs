use askama::Template;

// Truncated conditional expressions

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if }}")]
struct StopAfterIf;

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if ðŸ¦€ }}")]
struct OtherTokenAfterIf;

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if 3 + }}")]
struct EndInBinaryStateInCondition;

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 }}")]
struct ElseIsNotOptional; // actually valid in Jinja, defaults to `else 0`

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 else }}")]
struct StopAfterElse;

#[derive(Template)]
#[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 else ðŸ¦€ }}")]
struct OtherTokenAfterElse;

// Raw identifiers and strings are not keywords

#[derive(Template)]
#[template(ext = "html", source = "{{ then r#if condition else otherwise }}")]
struct RawIf;

#[derive(Template)]
#[template(ext = "html", source = "{{ then if condition r#else otherwise }}")]
struct RawElse;

#[derive(Template)]
#[template(ext = "html", source = r#"{{ then "if" condition else otherwise }}"#)]
struct StringIf;

#[derive(Template)]
#[template(ext = "html", source = r#"{{ then if condition "else" otherwise }}"#)]
struct StringElse;

fn main() {}
