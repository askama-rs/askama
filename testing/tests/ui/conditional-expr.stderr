error: conditional expression (`.. if .. else ..`) expects an expression after the keyword `if`
 --> <source attribute>:1:9
       "if }}"
 --> tests/ui/conditional-expr.rs:6:35
  |
6 | #[template(ext = "html", source = "{{ 1 + 2 if }}")]
  |                                   ^^^^^^^^^^^^^^^^

error: conditional expression (`.. if .. else ..`) expects an expression after the keyword `if`
 --> <source attribute>:1:9
       "if ðŸ¦€ }}"
  --> tests/ui/conditional-expr.rs:10:35
   |
10 | #[template(ext = "html", source = "{{ 1 + 2 if ðŸ¦€ }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^

error: in Askama, the else-case of a conditional expression (`.. if .. else ..`) is not optional
 --> <source attribute>:1:12
       "3 + }}"
  --> tests/ui/conditional-expr.rs:14:35
   |
14 | #[template(ext = "html", source = "{{ 1 + 2 if 3 + }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^

error: in Askama, the else-case of a conditional expression (`.. if .. else ..`) is not optional
 --> <source attribute>:1:12
       "3 + 4 }}"
  --> tests/ui/conditional-expr.rs:18:35
   |
18 | #[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^

error: conditional expression (`.. if .. else ..`) expects an expression after the keyword `else`
 --> <source attribute>:1:18
       "else }}"
  --> tests/ui/conditional-expr.rs:22:35
   |
22 | #[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 else }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: conditional expression (`.. if .. else ..`) expects an expression after the keyword `else`
 --> <source attribute>:1:18
       "else ðŸ¦€ }}"
  --> tests/ui/conditional-expr.rs:26:35
   |
26 | #[template(ext = "html", source = "{{ 1 + 2 if 3 + 4 else ðŸ¦€ }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unknown node `r`
 --> <source attribute>:1:8
       "r#if condition else otherwise }}"
  --> tests/ui/conditional-expr.rs:32:35
   |
32 | #[template(ext = "html", source = "{{ then r#if condition else otherwise }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: in Askama, the else-case of a conditional expression (`.. if .. else ..`) is not optional
 --> <source attribute>:1:11
       "condition r#else otherwise }}"
  --> tests/ui/conditional-expr.rs:36:35
   |
36 | #[template(ext = "html", source = "{{ then if condition r#else otherwise }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: failed to parse template source
 --> <source attribute>:1:8
       "\"if\" condition else otherwise }}"
  --> tests/ui/conditional-expr.rs:40:35
   |
40 | #[template(ext = "html", source = r#"{{ then "if" condition else otherwise }}"#)]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: in Askama, the else-case of a conditional expression (`.. if .. else ..`) is not optional
 --> <source attribute>:1:11
       "condition \"else\" otherwise }}"
  --> tests/ui/conditional-expr.rs:44:35
   |
44 | #[template(ext = "html", source = r#"{{ then if condition "else" otherwise }}"#)]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
